x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportCreateEnrollmentTokenAction.java

```diff
@@ -0,0 +1,111 @@
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.json.JsonXContent;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.http.HttpInfo;
+import org.elasticsearch.node.NodeService;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.security.SecurityContext;
+import org.elasticsearch.xpack.core.security.action.enrollment.CreateEnrollmentTokenAction;
+import org.elasticsearch.xpack.core.security.action.enrollment.CreateEnrollmentTokenRequest;
+import org.elasticsearch.xpack.core.security.action.enrollment.CreateEnrollmentTokenResponse;
+import org.elasticsearch.xpack.core.ssl.CertParsingUtils;
+import org.elasticsearch.xpack.security.authc.ApiKeyService;
+import org.elasticsearch.xpack.security.authc.support.ApiKeyGenerator;
+import org.elasticsearch.xpack.security.authz.store.CompositeRolesStore;
+
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.security.cert.X509Certificate;
+import java.util.Base64;
+import java.util.List;
+
+/**
+ * Transport action responsible for creating an enrollment token based on a request.
+ */
+
+public class TransportCreateEnrollmentTokenAction
+    extends HandledTransportAction<CreateEnrollmentTokenRequest, CreateEnrollmentTokenResponse> {
+
+    private final ApiKeyGenerator generator;
+    private final SecurityContext securityContext;
+    private final Environment environment;
+    private final NodeService nodeService;
+
+    @Inject
+    public TransportCreateEnrollmentTokenAction(TransportService transportService, ActionFilters actionFilters, ApiKeyService apiKeyService,
+                                                SecurityContext context, CompositeRolesStore rolesStore,
+                                                NamedXContentRegistry xContentRegistry, Environment environment, NodeService nodeService) {
+        super(CreateEnrollmentTokenAction.NAME, transportService, actionFilters, CreateEnrollmentTokenRequest::new);
+        this.generator = new ApiKeyGenerator(apiKeyService, rolesStore, xContentRegistry);
+        this.securityContext = context;
+        this.environment = environment;
+        this.nodeService = nodeService;
+    }
+
+    @Override
+    protected void doExecute(Task task, CreateEnrollmentTokenRequest request,
+                             ActionListener<CreateEnrollmentTokenResponse> listener) {
+        createEnrolmentToken(request, listener);
+    }
+
+    private void createEnrolmentToken(CreateEnrollmentTokenRequest request, ActionListener<CreateEnrollmentTokenResponse> listener) {
+        try {
+            generator.generateApiKeyForEnrollment(securityContext.getAuthentication(), request,
+                ActionListener.wrap(
+                    CreateApiKeyResponse -> {;
+                        final String httpCaCert = "httpCa.pem";
```

Comment:
I Think I had a change of heart around this and have an idea, let's chat on Monday morning.

Reply:
Leaving it as it is for now as we discussed

Reply:
See https://github.com/elastic/elasticsearch/pull/72207/files#diff-273a8e130ec3385b11fbcccb45472b75bccd1dd11a04175a3722bc05fc84be65R67 for an appoach as to how you can get the CA certificate from the keystore we expect to be used, without hardcoding the path/filename